cpp/src/interop/arrow_data_structures.cpp

```diff
@@ -359,15 +341,38 @@ arrow_table::arrow_table(ArrowSchema&& schema,
 {
   ArrowDeviceArray arr{.array = {}, .device_id = -1, .device_type = ARROW_DEVICE_CPU};
   ArrowArrayMove(&input, &arr.array);
-  container = arrow_table(std::move(schema), std::move(arr), stream, mr).container;
+  auto tmp     = arrow_table(std::move(schema), std::move(arr), stream, mr);
+  container    = tmp.container;
+  view_columns = std::move(tmp.view_columns);
+  cached_view  = tmp.cached_view;
 }
 
 arrow_table::arrow_table(ArrowArrayStream&& input,
                          rmm::cuda_stream_view stream,
                          rmm::device_async_resource_ref mr)
 {
-  auto tbl  = from_arrow_stream(&input, stream, mr);
-  auto tmp  = arrow_table(std::move(*tbl), get_table_metadata(tbl->view()), stream, mr);
-  container = tmp.container;
+  auto tbl     = from_arrow_stream(&input, stream, mr);
+  auto tmp     = arrow_table(std::move(*tbl), get_table_metadata(tbl->view()), stream, mr);
+  container    = tmp.container;
+  view_columns = std::move(tmp.view_columns);
+  cached_view  = tmp.cached_view;
+}
+
+void arrow_table::to_arrow_schema(ArrowSchema* output,
+                                  rmm::cuda_stream_view stream,
+                                  rmm::device_async_resource_ref mr) const
+{
+  ArrowSchemaDeepCopy(&container->schema, output);
+}
+
+void arrow_table::to_arrow(ArrowDeviceArray* output,
+                           ArrowDeviceType device_type,
+                           rmm::cuda_stream_view stream,
+                           rmm::device_async_resource_ref mr) const
+{
+  arrow_obj_to_arrow(*this, container, output, device_type, stream, mr);
 }
+
+table_view arrow_table::view() const { return cached_view; }
```

Comment:
have you considered creating the cached view on first use?

Reply:
I left a response here yesterday, but it looks like GH ate it :( I'm not sure why. Here it is again.

I did consider this, but I'm glad you ask because this is definitely one of the important considerations in this PR. The reason I chose to do this on construction is because it allows the `view` methods to be stream- and mr-free, matching the corresponding `table` method. Moreover, it ensures that the `view` methods are predictably cheap, also matching the expectations set by `table::view`. It also ensures that the member variables that store the underlying state do not need to be (de facto or de jure) optionals since they will be initialized when the object is, which avoids at least one possible error-prone pattern.

Reply:
Sounds like you did consider it :) 
Thanks for the details. The reasoning makes sense to me 👍

Reply:
Cool I'll resolve this thread then :) Anything else you'd like to see in this PR?

Reply:
Yes, I agree that this design is reasonable.